s3stream/src/main/java/com/automq/stream/s3/index/LocalStreamRangeIndexCache.java

```diff
@@ -0,0 +1,312 @@
+                    }
+                    cf.complete(null);
+                });
+        } finally {
+            readLock.unlock();
+        }
+        return cf;
+    }
+
+    public void init(int nodeId, ObjectStorage objectStorage) {
+        writeLock.lock();
+        try {
+            this.nodeId = nodeId;
+            this.objectStorage = objectStorage;
+            this.objectStorage.read(ObjectStorage.ReadOptions.DEFAULT, ObjectUtils.genIndexKey(0, nodeId))
+                .whenComplete((data, ex) -> {
+                    if (ex != null) {
+                        // cache not found
+                        initCf.complete(null);
+                        LOGGER.info("Sparse index not found for node {}", nodeId);
+                        return;
+                    }
+                    writeLock.lock();
+                    try {
+                        for (Map.Entry<Long, List<RangeIndex>> entry : LocalStreamRangeIndexCache.fromBuffer(data).entrySet()) {
+                            this.streamRangeIndexMap.put(entry.getKey(), new SparseRangeIndex(COMPACT_NUM, SPARSE_PADDING, entry.getValue()));
+                        }
+                    } finally {
+                        writeLock.unlock();
+                    }
+                    data.release();
+                    initCf.complete(null);
+                    LOGGER.info("Loaded sparse index from object storage for {} streams at node {}", streamRangeIndexMap.size(), nodeId);
+                });
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private <T> CompletableFuture<T> exec(Callable<T> r) {
+        return initCf.thenApply(v -> {
+            try {
+                return r.call();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    public void clear() {
+        writeLock.lock();
+        try {
+            streamRangeIndexMap.clear();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    public long nodeId() {
+        readLock.lock();
+        try {
+            return nodeId;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    public CompletableFuture<Void> append(Map<Long, RangeIndex> rangeIndexMap) {
+        return exec(() -> {
+            writeLock.lock();
+            try {
+                for (Map.Entry<Long, RangeIndex> entry : rangeIndexMap.entrySet()) {
+                    long streamId = entry.getKey();
+                    RangeIndex rangeIndex = entry.getValue();
+                    streamRangeIndexMap.computeIfAbsent(streamId,
+                        k -> new SparseRangeIndex(COMPACT_NUM, SPARSE_PADDING)).append(rangeIndex);
+                }
+            } finally {
+                writeLock.unlock();
+            }
+            return null;
+        });
+    }
+
+    public CompletableFuture<Void> compact(Map<Long, RangeIndex> rangeIndexMap, Set<Long> compactedObjectIds) {
```

Comment:
When to clear the sparse index of a stream.

Reply:
solved in #1587